projects/oit_demo/DepthPeeling.cpp

```diff
@@ -0,0 +1,289 @@
+        gpCreateInfo.polygonMode                        = grfx::POLYGON_MODE_FILL;
+        gpCreateInfo.cullMode                           = grfx::CULL_MODE_NONE;
+        gpCreateInfo.frontFace                          = grfx::FRONT_FACE_CCW;
+        gpCreateInfo.depthReadEnable                    = true;
+        gpCreateInfo.depthWriteEnable                   = true;
+        gpCreateInfo.blendModes[0]                      = grfx::BLEND_MODE_NONE;
+        gpCreateInfo.outputState.renderTargetCount      = 1;
+        gpCreateInfo.outputState.renderTargetFormats[0] = mDepthPeeling.layerTextures[0]->GetImageFormat();
+        gpCreateInfo.outputState.depthStencilFormat     = mDepthPeeling.depthTextures[0]->GetImageFormat();
+        gpCreateInfo.pPipelineInterface                 = mDepthPeeling.layerPipelineInterface;
+
+        grfx::ShaderModulePtr VS, PS;
+
+        PPX_CHECKED_CALL(CreateShader("oit_demo/shaders", "DepthPeelingLayer_First.vs", &VS));
+        PPX_CHECKED_CALL(CreateShader("oit_demo/shaders", "DepthPeelingLayer_First.ps", &PS));
+        gpCreateInfo.VS = {VS, "vsmain"};
+        gpCreateInfo.PS = {PS, "psmain"};
+        PPX_CHECKED_CALL(GetDevice()->CreateGraphicsPipeline(&gpCreateInfo, &mDepthPeeling.layerPipeline_FirstLayer));
+        GetDevice()->DestroyShaderModule(VS);
+        GetDevice()->DestroyShaderModule(PS);
+
+        PPX_CHECKED_CALL(CreateShader("oit_demo/shaders", "DepthPeelingLayer_Others.vs", &VS));
+        PPX_CHECKED_CALL(CreateShader("oit_demo/shaders", "DepthPeelingLayer_Others.ps", &PS));
+        gpCreateInfo.VS = {VS, "vsmain"};
+        gpCreateInfo.PS = {PS, "psmain"};
+        PPX_CHECKED_CALL(GetDevice()->CreateGraphicsPipeline(&gpCreateInfo, &mDepthPeeling.layerPipeline_OtherLayers));
+        GetDevice()->DestroyShaderModule(VS);
+        GetDevice()->DestroyShaderModule(PS);
+    }
+
+    ////////////////////////////////////////
+    // Combine
+    ////////////////////////////////////////
+
+    // Descriptor
+    {
+        grfx::DescriptorSetLayoutCreateInfo layoutCreateInfo = {};
+        layoutCreateInfo.bindings.push_back(grfx::DescriptorBinding{SHADER_GLOBALS_REGISTER, grfx::DESCRIPTOR_TYPE_UNIFORM_BUFFER, 1, grfx::SHADER_STAGE_ALL_GRAPHICS});
+        layoutCreateInfo.bindings.push_back(grfx::DescriptorBinding{CUSTOM_SAMPLER_0_REGISTER, grfx::DESCRIPTOR_TYPE_SAMPLER, 1, grfx::SHADER_STAGE_ALL_GRAPHICS});
+        layoutCreateInfo.bindings.push_back(grfx::DescriptorBinding{CUSTOM_TEXTURE_0_REGISTER, grfx::DESCRIPTOR_TYPE_SAMPLED_IMAGE, DEPTH_PEELING_LAYERS_COUNT, grfx::SHADER_STAGE_ALL_GRAPHICS});
+        PPX_CHECKED_CALL(GetDevice()->CreateDescriptorSetLayout(&layoutCreateInfo, &mDepthPeeling.combineDescriptorSetLayout));
+
+        PPX_CHECKED_CALL(GetDevice()->AllocateDescriptorSet(mDescriptorPool, mDepthPeeling.combineDescriptorSetLayout, &mDepthPeeling.combineDescriptorSet));
+
+        grfx::WriteDescriptor writes[10] = {};
```

Comment:
Hum, yeah it sucks. I can't find a satisfying way to do that in a way that would break compilation if not done properly.
"DEPTH_PEELING_LAYER_COUNT + 2" or "10" ... I feel both solutions are afflicted with the same problem... that crappy constant that comes from nowhere.
Unless there is a way to automatically express the number of resources to bind, I fear it will never be satisfying.
Thankfully, the validation layer usually catch most, if not all errors, around descriptors (at least it caught mine :) ).

Reply:
Yeah. on embedded, magics are a pain, but it's common to have stuff like:
`SOME_PCI_REGISTER_0 + 1` to mix weird magics with index based loops.
Maybe something like
`DEPTH_PEELING_TEXTURE_0 + i`? and the other 2 indices defined?
It doesn't prevent bugs, but at least makes it easier to catch issues as I think a completely different name is more visible than a random integer?

Reply:
So, actually, my register numbers are not contiguous. :(
For example, CUSTOM_TEXTURE_0_REGISTER (the first depth peeling texture) is equal to 3, not 2.
Let me use your first suggestion (DEPTH_PEELING_LAYER_COUNT + 2), which should at least document it a bit.
I will add it in my next PR.
I realized I should probably also not hardcode the value in the call to UpdateDescriptors... I will fix that for all algorithms as well in my next PR.